[WTF] Remove pthread_main_np use in MainThreadGeneric.

jameshilliard opened the pull request:
#### 799dd04c3edbd3352761e2559165f1e514384ad5
<pre>
[WTF] Remove pthread_main_np use in MainThreadGeneric.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244153">https://bugs.webkit.org/show_bug.cgi?id=244153</a>

Reviewed by Michael Catanzaro.

It was determined in <a href="https://bugs.webkit.org/show_bug.cgi?id=243401">https://bugs.webkit.org/show_bug.cgi?id=243401</a> that
isMainThread() can not simply check if the current thread is the initial
process thread but must compare the current thread against the thread
set when initializeMainThreadPlatform is called to handle cases when
WebKit is run under applications like Eclipse which don&apos;t initialize
WebKit in the main process thread.

As such we should not use pthread_main_np() which compares the current
thread against the initial process thread.

* Source/WTF/wtf/generic/MainThreadGeneric.cpp:
(WTF::initializeMainThreadPlatform):
(WTF::isMainThread):

Canonical link: <a href="https://commits.webkit.org/253647@main">https://commits.webkit.org/253647@main</a>
</pre>









<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e764185a71f348422efcc55a9648faabd987a9eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86678 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30737 "Built successfully") | [✅ ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/17608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95521 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149252 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29122 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25534 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78859 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90763 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92294 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23522 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73598 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23585 "Passed tests") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/17608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/66592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78613 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26893 "Built successfully") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/17608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72250 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26812 "Built successfully") | [✅ ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/17608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25798 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2591 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28491 "Built successfully") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/36601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75033 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28435 "Built successfully") | [✅ ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/17608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16598 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode") | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WTF/wtf/generic/MainThreadGeneric.cpp:
```diff
@@ -32,32 +32,21 @@
 
 bool isMainThread()
 {
-#if HAVE(PTHREAD_MAIN_NP)
-    return pthread_main_np();
-#else
     return pthread_equal(pthread_self(), mainThread);
-#endif
```
Note that macOS uses `pthread_main_np()` here. I don't think that the title of this PR accurate - it may do what's necessary on Linux, but it doesn't align it with other platforms.

On macOS and iOS, WebKit2 APIs must be called from the main thread (same as UI thread, same as what pthread_main_np considers main). Similarly, there is a number of platforms APIs that are main thread only, so WebKit hops to the main thread to use them. So the Eclipse use case described here would be invalid for us.

jameshilliard replied:
> I don't think that the title of this PR accurate - it may do what's necessary on Linux, but it doesn't align it with other platforms.

AFAIU it's not a Linux specific issue so much as an issue dependent on how one uses webkit(ie if one launches it via a non-process main thread with GTK).

> same as UI thread

I think this is still the case on Linux, AFAIU main thread is still always the UI thread, it's just that both may not be the process main thread in some cases.

> On macOS and iOS, WebKit2 APIs must be called from the main thread (same as UI thread, same as what pthread_main_np considers main). Similarly, there is a number of platforms APIs that are main thread only, so WebKit hops to the main thread to use them. So the Eclipse use case described here would be invalid for us.

Hmm, what API's are those? Maybe we need something with different semantics for those cases like a `isProcessMainThread()`, that way other platforms like Linux could also make use of that for cases where one actually wants to see if one is running on the main process thread as opposed to just the thread which initialized webkit.

aproskuryakov replied:
APIs in `Source/WebKit/UIProcess/API/Cocoa/` are currently only to be called from the main thread. So, initialization always happens on the main thread.

It's different from JavaScriptCore, which is usable from secondary threads.

Again, I'm _not_ saying that this patch is wrong. But it's not making semantics the same across all platforms.

mcatanzaro replied:
I'd say this change to MainThreadGeneric is surely good, but the commit message is indeed not right. You're unifying the semantics of isMainThread for non-Cocoa Unix platforms in a way that will fix applications that would currently be broken on non-Linux platforms, so that BSD users have a chance to use applications that depend on this. That's good. But the semantics for Cocoa are still different.

I wouldn't try to change the Cocoa implementation to match MainThreadGeneric; regardless of whether the change would be correct on Apple platforms or not, evidently no applications need it or they would have complained by now. So I would leave the code exactly as you've done in this pull request, and instead just come up with a different commit message. That way we don't need to figure out whether it's OK for the WebKit main thread to not be the same as the operating system main thread on Apple platforms. Maybe you could add some comments to explain things, though, since this is very subtle and would be easy to break again in the future. Your original attempt to change this seemed pretty reasonable, after all, and someone might well try again at some point if we don't leave a warning.

jameshilliard replied:
> I'd say this change to MainThreadGeneric is surely good, but the commit message is indeed not right. You're unifying the semantics of isMainThread for non-Cocoa Unix platforms in a way that will fix applications that would currently be broken on non-Linux platforms, so that BSD users have a chance to use applications that depend on this. That's good. But the semantics for Cocoa are still different.

I reworded it a little, is that better?

> Maybe you could add some comments to explain things, though, since this is very subtle and would be easy to break again in the future. Your original attempt to change this seemed pretty reasonable, after all, and someone might well try again at some point if we don't leave a warning.

Added a comment so that it's clear that `mainThread` may not be the process initial thread.

mcatanzaro replied:
> I reworded it a little, is that better?

No, because the behavior is not the same across all operating systems. :P  I would say something like "[WTF] Remove pthread_main_np use in MainThreadGeneric."

jameshilliard replied:
> No, because the behavior is not the same across all operating systems. :P

Well I was referring to the implementation behavior of just `MainThreadGeneric` itself(ie eliminating the differing codepaths for `HAVE(PTHREAD_MAIN_NP)`).

> I would say something like "[WTF] Remove pthread_main_np use in MainThreadGeneric."

Changed to that.

mcatanzaro commented on Source/WTF/wtf/generic/MainThreadGeneric.cpp:
```diff
@@ -32,32 +32,22 @@
 void initializeMainThreadPlatform()
 {
-#if !HAVE(PTHREAD_MAIN_NP)
+    // mainThread is often but not always the initial process thread
```
We have a code style rule that comments are sentences and should end in periods:

```suggestion
    // mainThread is often but not always the initial process thread.
```

But I would go even farther here to describe things a bit more:

```suggestion
    // WebKit APIs must be consistently used from exactly one thread. The thread that initializes WebKit
    // is considered the "WebKit main thread," and it is an error to use WebKit APIs from any other thread.
    // The WebKit main thread need not be the application's actual OS-level main thread, which might be
    // controlled by a language runtime or virtual machine; for example, in Eclipse, the OS main thread is
    // controlled by the JVM, while the separate WebKit main thread runs all the GUI and WebKit stuff.
```

This is different from Apple systems, but that's fine.

jameshilliard committed:
[WTF] Remove pthread_main_np use in MainThreadGeneric.
https://bugs.webkit.org/show_bug.cgi?id=244153

Reviewed by Michael Catanzaro.

It was determined in https://bugs.webkit.org/show_bug.cgi?id=243401 that
isMainThread() can not simply check if the current thread is the initial
process thread but must compare the current thread against the thread
set when initializeMainThreadPlatform is called to handle cases when
WebKit is run under applications like Eclipse which don't initialize
WebKit in the main process thread.

As such we should not use pthread_main_np() which compares the current
thread against the initial process thread.

* Source/WTF/wtf/generic/MainThreadGeneric.cpp:
(WTF::initializeMainThreadPlatform):
(WTF::isMainThread):

Canonical link: https://commits.webkit.org/253647@main
